File: windows/tests/RealtimeTranslator.Core.Tests/AppReleaseVersionTests.cs
using System.Reflection;
using RealtimeTranslator.Core.Settings;
using Xunit;

namespace RealtimeTranslator.Core.Tests;

public sealed class AppReleaseVersionTests
{
    // Given: 空や空白だけの入力
    // When: 表示値へ正規化する
    // Then: 未リリースの 0.0.0 を返す
    [Theory]
    [InlineData(null)]
    [InlineData("")]
    [InlineData("   ")]
    [InlineData("+build")]
    public void BlankRawBecomesUnpublished(string? raw) =>
        Assert.Equal(AppReleaseVersion.Unpublished, AppReleaseVersion.DisplayValue(raw));

    // Given: リリースタグやアセンブリ InformationalVersion
    // When: 表示値へ正規化する
    // Then: 先頭の v と + 以降のビルドメタデータを落とし、本文だけ残す
    [Theory]
    [InlineData("0.1.0", "0.1.0")]
    [InlineData("v0.1.0", "0.1.0")]
    [InlineData("V0.1.0", "0.1.0")]
    [InlineData("  v0.1.0  ", "0.1.0")]
    [InlineData("0.1.0+abc123", "0.1.0")]
    [InlineData("v0.1.0-rc.1+sha", "0.1.0-rc.1")]
    [InlineData("pr12", "pr12")]
    [InlineData("very", "very")]
    public void DisplayValueStripsTagPrefixAndBuildMetadata(string raw, string expected) =>
        Assert.Equal(expected, AppReleaseVersion.DisplayValue(raw));

    // Given: 未リリースの既定 Version (Directory.Build.props)
    // When: Core アセンブリの InformationalVersion を読む
    // Then: 設定画面と同じ 0.0.0 になる
    [Fact]
    public void UnpublishedAssemblyVersionIsZero()
    {
        var informational = typeof(AppReleaseVersion).Assembly
            .GetCustomAttribute<AssemblyInformationalVersionAttribute>()
            ?.InformationalVersion;

        Assert.Equal(AppReleaseVersion.Unpublished, AppReleaseVersion.DisplayValue(informational));
    }
}

File: RealtimeTranslator/Tests/AppReleaseVersionTests.swift
import XCTest
@testable import RealtimeTranslator

final class AppReleaseVersionTests: XCTestCase {
    // Given: 空や空白だけの入力
    // When: 表示値へ正規化する
    // Then: 未リリースの 0.0.0 を返す
    func testBlankRawBecomesUnpublished() {
        XCTAssertEqual(AppReleaseVersion.displayValue(from: nil), AppReleaseVersion.unpublished)
        XCTAssertEqual(AppReleaseVersion.displayValue(from: ""), AppReleaseVersion.unpublished)
        XCTAssertEqual(AppReleaseVersion.displayValue(from: "   "), AppReleaseVersion.unpublished)
        XCTAssertEqual(AppReleaseVersion.displayValue(from: "+build"), AppReleaseVersion.unpublished)
    }

    // Given: リリースタグや InformationalVersion
    // When: 表示値へ正規化する
    // Then: 先頭の v と + 以降のビルドメタデータを落とし、本文だけ残す
    func testDisplayValueStripsTagPrefixAndBuildMetadata() {
        XCTAssertEqual(AppReleaseVersion.displayValue(from: "0.1.0"), "0.1.0")
        XCTAssertEqual(AppReleaseVersion.displayValue(from: "v0.1.0"), "0.1.0")
        XCTAssertEqual(AppReleaseVersion.displayValue(from: "V0.1.0"), "0.1.0")
        XCTAssertEqual(AppReleaseVersion.displayValue(from: "  v0.1.0  "), "0.1.0")
        XCTAssertEqual(AppReleaseVersion.displayValue(from: "0.1.0+abc123"), "0.1.0")
        XCTAssertEqual(AppReleaseVersion.displayValue(from: "v0.1.0-rc.1+sha"), "0.1.0-rc.1")
        XCTAssertEqual(AppReleaseVersion.displayValue(from: "pr12"), "pr12")
        XCTAssertEqual(AppReleaseVersion.displayValue(from: "very"), "very")
    }

    // Given: テストホストの Info.plist
    // When: current を読む
    // Then: CFBundleShortVersionString と同じ正規化結果になる
    func testCurrentMatchesBundleShortVersionString() {
        let raw = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
        XCTAssertEqual(AppReleaseVersion.current, AppReleaseVersion.displayValue(from: raw))
        XCTAssertFalse(AppReleaseVersion.current.isEmpty)
    }
}
